### ISSUE TITLE
Change the order of things checked for languages

### ISSUE THREAD

**Issue by pbrisbin** [yesodweb/yesod#1720]
We've been moving to a more mature system of localization, and as part of that we are setting our users' language in their Session at log-in, so our Frontend client can just trust API responses are correctly localized in the session. This has been really great.

Sometimes we need a particular API response in a specific language for just that request. However, overriding it via the URL doesn't work because of the order things are checked in:

> - The _LANG user session variable.
> - The _LANG get parameter.
> - The _LANG cookie.
> - Accept-Language HTTP header.

[Source](https://www.stackage.org/haddock/lts-17.9/yesod-core-1.6.18.8/Yesod-Core-Handler.html#v:languages).

This doesn't work for our use case, but I think it's objectively surprising and a borderline bug. I would expect properties of _this request_ to be checked ahead of properties of _this session_, not the reverse. I think it would make more sense to "fall back" to something ambient (session, cookie) only if something more explicit (param, header) isn't given.

So, I suggest:

- Accept-Language HTTP header.
- The _LANG get parameter.
- The _LANG cookie.
- The _LANG user session variable.

As the order to check.

I could totally be missing a reason for the current order, and the behavior change is subtle enough that it could really break in-the-wilds apps in hard-to-debug-ways, so I wonder: if I opened a PR to change this, would it be accepted?

**Comment by snoyberg** [comment]
The accept-language header should IMO definitely end up at the end of the list. Basically every browser will send it by default, and if the user selects a different language than what their browser defaults to, we need to respect that.

I think we're both in agreement that the get parameter should trump the cookie, so no issue there.

I'm surprised that you think the _LANG cookie should trump the user session variable, or at the very least that this is something worth worrying about. These seem to be at the same level of "per session"-ness. What's the reason here?

The user session variable vs get parameter (and cookie for that matter) are an interesting case. I think the reason I set it up this way in the first place is that, if someone copy-pastes a URL and sends it to another user, we'd like it to show up in the language that they're expecting. But I can definitely see your point here, and I'd be open to making the get parameter supersede anything else.

In other words, IMO, swapping the current preference between the first two would be the best step forward.

**Comment by pbrisbin** [comment]
Ah, moving up Accept-Language was a total mistake. I have been bitten before by the automatic browser-set values, and we basically consider it useless. I don't know why I forgot that in this moment. For some reason, I was thinking of it as a standards-backed place for a client to put a language, and one that supported a ranked choices list, so "better" than a functionally-equivalent GET parameter. Oops.

For clarification of my reasoning on the others (particularly Cookie), I was looking at "ambient" values vs "explicit" values. As an analogy, using the Session might be like a command-line tool respecting `~/.config/my-cli/language` and using the query parameter is calling `my-cli --language x`. We'd all expect the command-line flag to win. And we agree about considering query parameter like this, it sounds like. My choice to let the Cookie override was also because (conceptually) it's an attribute of the request, and so should override the ambient Session value. This was semi-arbitrary though, and possibly a stretch. Practically speaking, Session and Cookie are equivalent for us (as is likely the case for most Yesod users), making it more of an ambient kind of storage, even if technically it's a value explicit in every request.

> if someone copy-pastes a URL and sends it to another user, we'd like it to show up in the language that they're expecting

Oh, that is an interesting trade-off! On balance, I think I still feel the parameter winning is less surprising even if it means this use-case is unsupported.

So, your suggestion:

- The _LANG get parameter.
- The _LANG user session variable.
- The _LANG cookie.
- Accept-Language HTTP header.

SG(great)TM!

**Comment by snoyberg** [comment]
Yes, let's do it.

**Comment by pbrisbin** [comment]
I'll try to work up the PR this week. Considering the (IMO) subtle and possibly hair-pulling breakage it could cause for this behavior to change, should I consider it a major version bump? Or do you think we don't worry about it and do it as minor (or even patch) since no typed contracts are changing?

**Comment by snoyberg** [comment]
If we're really worried about the breakage, I think the right thing to do is to feature gate the change behind a method in the `Yesod` typeclass.

**Comment by pbrisbin** [comment]
That's a good idea, if we're worried. Are you worried about it? I'm a soft-no, personally, and would just do it in minor. I said originally it's a borderline bugfix, IMO.

**Comment by pbrisbin** [comment]
Forgot to close this when the PR landed. Oops.
